man/man1/tangoktl.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH TANGOKTL 1 "July 1, 2025" "Tango tools" "User Commands"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7) and groff_man(7)
.\" .SH        section heading
.\" .SS        secondary section heading
.\"
.\"
.\" To preview this page as plain text: nroff -man tangoktl.1
.\"
.SH NAME

tangoktl \- Tool for Reading Tango Devices running in a Kubernetes cluster.

.SH SYNOPSIS

.B tangoktl
.RI [ options ]

.SH DESCRIPTION

Tangoktl reads information fields, attributes, command and property configuration and
values of all Tango devices with a single command. The ouput format is configurable. It
can be called from bash or Python. It can be configured with a JSON file. Regular
expression filtering of Tango entities is supported. Support for Tango devices running
in a Kubernetes cluster is built in.

For a summary on how to use \fBtangoktl\fR, see 'tangoktl \-\-help' for all available command\(hyline options.

.SH OPTIONS

When reading Tango devices, the output can be in plain text, YAML, HTML, JSON or Markdown format.

By default, devices with names that start with _sys_ or _dserver_ are not shown. This
can be overridden using a command line flag or in the configuration file. The Tango
devices can be filtered by attribute, command, property or class name. Matches can be
done on full or partial matches.  To shorten the running time, there is an option to
limit output to one device per class.

This enhanced version of **tangoctl** has support for Tango devices
running in a Kubernetes cluster. The namespaces in the cluster can be listed. By setting
a standard pod name for Tango database servers in the configuration file, the FQDN of
the server can be derived and the IP address can be looked up.

By default, commands are not run since it might result in devices ending up in unwanted
or unsafe states. The configuration file contain commands that are safe to be read,
for example Status.

There is a basic test function. Values for attributes, commands or properties are
specified in a JSON file. These are written and then read. The test fails if the
expected value is not returned.

Note that values for device, attribute, command or property are not case sensitive.
Partial matches for strings longer than 4 charaters are OK.

Run the following commands where applicable:
QueryClass,QueryDevice,QuerySubDevice,GetVersionInfo,State,Status

Run commands with device name as parameter where applicable:
DevLockStatus,DevPollStatus,GetLoggingTarget

.TP
.B -a, --show-attribute
flag for reading attributes

.TP
.B -b, --show-pod
read pod names

.TP
.B -c, --show-command
flag for reading commands

.TP
.B -d, --show-dev
list Tango device names

.TP
.B -e, --everything
show all devices - do not skip sys, dserver

.TP
.B --exact
show only exact matches for names

.TP
.B -f, --full, --large
display in full

.TP
.B -i, --show-db
display hostname and IP address of Tango host

.TP
.B -j, --json
output in JSON format

.TP
.B -k, --show-class
list Tango device classes

.TP
.B -l, --list
display device name, status and values

.TP
.B -n, --show-ns
read Kubernetes namespaces

.TP
.B --pod-df
read pod file system space usage

.TP
.B --pod-domain
read pod domain name

.TP
.B --pod-env
read pod environment variables

.TP
.B --pod-free
read pod free memory

.TP
.B --pod-host
read pod host name

.TP
.B --pod-mpstat
read pod processor related statistics

.TP
.B --pod-ps
read active processes in pod

.TP
.B --pod-top
read system summary information in pod

.TP
.B --pod-uptime
read how long pods have been running

.TP
.B -p, --show-property
flag for reading properties

.TP
.B --ping
ping device

.TP
.B -q, --quiet
do not display progress bars, set logging level to WARNING

.TP
.B -Q
do not display progress bars, set logging level to ERROR

.TP
.B -r, --reverse
reverse sort order

.TP
.B -s, --short, --small
display device name and status

.TP
.B -t, --txt
output in text format

.TP
.B --tree
display Tango device names as a tree

.TP
.B -u, --md
output in markdown format

.TP
.B --unique
only read one device for each class

.TP
.B -w, --html
output in HTML format

.TP
.B -y, --yaml
output in YAML format

.TP
.B -z, --show-svc
read service names

.TP
.B -0, --off
turn device off

.TP
.B -1, --on
turn device on

.TP
.B --simul=0,1
set simulation mode off or on

.TP
.B --admin=0,1
set admin mode off or on

.TP
.BI "-A " ATTRIBUTE ", --attribute=" ATTRIBUTE
attribute name, e.g. 'obsState' (not case sensitive)

.TP
.BI "-C " COMMAND ", --command=" COMMAND
command name, e.g. 'Status' (not case sensitive)

.TP
.BI "-D " DEVICE ",--device=" DEVICE
device name, e.g. 'csp' (not case sensitive, only a part is needed)

.TP
.BI "-F " FILE ", --cfg=" FILE
override configuration from file

.TP
.BI "-H " HOST ", --host=" HOST
Tango database host and port, e.g. 10.8.13.15:10000

.TP
.BI "--indent=" INDENT
indentation for JSON output, default is 4

.TP
.BI "-I " FILE ", --input=FILE
input file name

.TP
.BI "-J " FILE ", --json-dir=" PATH
directory with JSON input file, e.g. 'resources'

.TP
.BI "-K " CLASS ", --class=" CLASS
Tango device class, e.g. 'MidCspSubarray' (not case sensitive)

.TP
.BI "-N " K8S_NS ", --namespace=" K8S_NS ", --ns=" K8S_NS
Kubernetes namespace for Tango database, e.g. 'integration'

.TP
.BI "-O " FILE ", --output=" FILE
output file name

.TP
.BI "-P " PROPERTY ", --property=" PROPERTY
property name, e.g. 'Model' (not case sensitive)

.TP
.BI "-R " PORT ", --port=" PORT
Tango database port, e.g. 10000

.TP
.BI "-W " VALUE ", --value=" VALUE
value for Tango attribute

.TP
.BI "-X " CONTEXT ", --cfg=" FILE
set Kubernetes context

.TP
.BI "-Z " TIMEOUT ", --timeout=" FILE
set Kubernetes context

.Sh "ENVIRONMENT VARIABLES"
.sp
.PP
\fITANGO_HOST\fR
.RS 4
This is the name of the host where the Tango database server is running, i.e.
.I tangoctl.json
or
.I tangoktl.json

.SH FILES

Tangoktl uses a configuration file named \fItangoktl.json\fR. It is stored in the same directory as the excutable.

There are no hard-coded attribute, command or property names. The columns in the list
are configured in this JSON file.

The test function. uses values for attributes, commands or properties that are
specified in JSON files.

.SH EXAMPLES

.TP
Display version number
tangoktl --version

.TP
Display help
\fBtangoktl --help

.TP
Display Kubernetes namespaces
\fBtangoktl --show-ns\fR

.TP
Display Tango database address in the namespace \fIintegration\fR
\fBtangoktl --show-db --namespace=integration\fR

.TP
Search for devices named \fItalon\fR in the namespace \fIintegration\fR
\fBtangoktl -N integration -D talon -l\fR
q
.TP
Display classes and Tango devices associated with them in the namespace \fIintegration\fR
\fBtangoktl -d -N integration\fR

.TP
List Tango device names in the namespace \fIintegration\fR
\fBtangoktl -l -N integration\fR

.TP
Display all Tango devices in YAML formal (will take a long time) in the namespace \fIintegration\fR
\fBtangoktl -N integration --yaml\fR

.TP
Filter on device \fIska_mid/tm_leaf_node/csp_subarray01\fR in the namespace \fIintegration\fR
\fBtangoktl -f -N integration -D ska_mid/tm_leaf_node/csp_subarray01\fR

.TP
Search for devices with an attribute named \fIpower\fR in the namespace \fItest-equipment\fR
\fBtangoktl -K test-equipment -A power\fR

.TP
Search for devices with a command named \fIstatus\fR in the namespace \fIintegration\fR
\fBtangoktl -l -N integration -C status\fR

.TP
Search for devices with a property named \fIpower\fR in the namespace \fIintegration\fR
\fBtangoktl -l -N integration -P power\fR

.SH Interfacing with other utilities

.TP
Display Tango device names in the namespace \fItest-equipment\fR in markdown format and pretty print it with \fBmdless\fR
\fB./src/ska_tangoctl/tango_kontrol/tangoktl.py -N test-equipment -macpq | mdless\fR

.TP
Read Tango device names in the namespace \fIintegration-dish-lmc-ska100\fR in YAML format and use \fByq\fR to extract their names
tangoktl.py -N integration-dish-lmc-ska100 -ya | yq  '.devices[] | .name'

.TP
Read Tango device names in the namespace \fIintegration-dish-lmc-ska100\fR in YAML format and use \fByq\fR to extract the name and status of each one
tangoktl.py -N integration-dish-lmc-ska100 -ya | jq '.devices[] | .name,.attributes.Status.data.value' integration-dish-lmc-ska100.json

.SH Testing Tango devices

.TP
Test a Tango device
\fBtangoktl.py [TANGODB] [TEST] -D DEVICE [--simul=0,1]\fR

.TP
Test a Tango device and read attributes
\fBtangoktl.py -a -N K8S_NS|-H HOST -D DEVICE [--simul=0,1]\fR

.TP
Display attribute and command names for a Tango device
\fBtangoktl.py -c -N K8S_NS|-H HOST -D DEVICE\fR

.TP
Turn a Tango device on
\fBtangoktl.py --on -N K8S_NS|-H HOST -D DEVICE [--simul=0,1]\fR

.TP
Turn a Tango device off
\fBtangoktl.py --off -N K8S_NS|-H HOST -D DEVICE [--simul=0,1]\fR

.TP
Set a Tango device to standby mode
\fBtangoktl.py --standby -N K8S_NS|-H HOST -D DEVICE [--simul=0,1]\fR

.TP
Change admin mode on a Tango device
\fBtangoktl.py --admin=<0|1>\fR

.TP
Display status of a Tango device
\fBtangoktl.py --status=0,1 -H HOST -D DEVICE\fR

.TP
Check events for attribute of a Tango device
\fBtangoktl.py -N staging-dish-lmc-ska001 -D DEVICE -A ATTRIBUTE\fR

.TP
Display tangoktl.py test input files
\fBtangoktl.py --json-dir=resources/\fR

.TP
Run test, reading from input file
\fBtangoktl.py --ns=staging-dish-lmc-ska001 --input=resources/dev_online.json\fR

.TP
Run test file:
\fBtangoktl.py [TANGODB] --in PATH -V

.SH "SEE ALSO"

\fBtangoctl\fR(1)

.SH REFERENCES

The full documentation for \fBtangoktl\fR is available at
\fBhttps://ska-telescope-ska-tangoktl.readthedocs.io/en/latest/\fR

An experimental C++ version with a subset of the above functionality is available at

\fBhttps://gitlab.com/jcoetzer/tangoctl_cpp\fR

.SH AUTHORS

This utility is written by the SKAO Development Team <https://skao.int>.


CHANGELOG.md
# Version History

## Unreleased

* [STS-654] create this repository

## 0.3.0

* [REL-1376] update documentation and tidy up command line options

## 0.3.1

* [REL-1381] Fix listing of devices with unique class option and installation; remove semver and slack-sdk

## 0.3.2

* [REL-1404] Fix documentation build on readthedocs.io

## 0.3.3

* [REL-1430] Add API documentation

## 0.3.4

* [REL-1435] Fix configuration

## 0.3.5

* [REL-1440] Update help message 

## 0.4.0

* [REL-1450] Add device tree and support for regular expressions in namespaces

## 0.4.1

* [REL-1454] Fix script names 

## 0.5.0

* [REL-1746] Add web interface

## 0.6.0

* [REL-2208] Various fixes and enhancements


src/ska_tangoctl/tango_kontrol/tangoktl.py
#!/usr/bin/env python
"""Read all information about Tango devices."""
import json
import logging
import os
import sys

import yaml

from ska_tangoctl import __version__
from ska_tangoctl.tango_control.disp_action import DispAction
from ska_tangoctl.tango_control.tango_database import TangoHostInfo, get_tango_hosts
from ska_tangoctl.tango_control.tango_device_tree import device_tree
from ska_tangoctl.tango_control.test_tango_device import TestTangoDevice
from ska_tangoctl.tango_kontrol.tango_kontrol import TangoKontrol
from ska_tangoctl.tla_jargon.tla_jargon import print_jargon

try:
    from ska_tangoctl.k8s_info.get_k8s_info import KubernetesInfo
except ModuleNotFoundError:
    KubernetesInfo = None  # type: ignore[assignment,misc]

logging.basicConfig(level=logging.WARNING, stream=sys.stderr)
_module_logger = logging.getLogger("tango_kontrol")
_module_logger.setLevel(logging.WARNING)


def main() -> int:  # noqa: C901
    """
    Read and display Tango devices.

    :return: error condition
    """
    k8s: KubernetesInfo = KubernetesInfo(_module_logger, None)
    tangoktl: TangoKontrol = TangoKontrol(
        _module_logger, k8s.context, k8s.cluster, k8s.domain_name
    )

    # Read command line options
    rc: int = tangoktl.read_command_line_k8s(sys.argv)
    if rc == 0:
        pass
    elif rc == 3:
        tangoktl.usage3(os.path.basename(sys.argv[0]))
        return 1
    elif rc == 4:
        tangoktl.usage4(os.path.basename(sys.argv[0]))
        return 1
    else:
        _module_logger.error("Read command line returned %d", rc)
        return 1

    # Read configuration
    tangoktl.read_config()

    _module_logger.info("Read Tango:\n%s", tangoktl)

    if tangoktl.disp_action.show_version:
        print(f"{os.path.basename(sys.argv[0])} version {__version__}")
        return 0

    if tangoktl.disp_action.show_jargon:
        print_jargon()
        return 0

    if tangoktl.disp_action.show_ctx:
        tangoktl.show_contexts()
        return 0

    if tangoktl.disp_action.show_ns:
        tangoktl.show_namespaces()
        return 0

    if tangoktl.disp_action.show_log and tangoktl.k8s_pod:
        tangoktl.show_pod_log()
        return 0

    if tangoktl.show_svc:
        tangoktl.show_services()
        return 0

    if tangoktl.disp_action.show_pod and tangoktl.disp_action.check(DispAction.TANGOCTL_TXT):
        tangoktl.set_output()
        tangoktl.list_pod_names(tangoktl.k8s_ns)
        tangoktl.unset_output()
        return 0

    if tangoktl.disp_action.show_proc and tangoktl.k8s_pod:
        tangoktl.print_pod_procs()
        return 0

    if tangoktl.json_dir:
        tangoktl.read_input_files(tangoktl.json_dir)
        return 0

    if tangoktl.pod_cmd and tangoktl.k8s_pod:
        tangoktl.show_pod(tangoktl.pod_cmd)
        return 0

    if tangoktl.pod_cmd:
        tangoktl.show_pods(tangoktl.pod_cmd)
        return 0

    if tangoktl.disp_action.show_pod and tangoktl.k8s_pod:
        pod = k8s.get_pod_desc(tangoktl.k8s_ns, tangoktl.k8s_pod)
        if not tangoktl.disp_action.indent:
            tangoktl.disp_action.indent = 4
        tangoktl.set_output()
        if tangoktl.disp_action.check(DispAction.TANGOCTL_YAML):
            print(
                yaml.dump(pod.to_dict(), indent=tangoktl.disp_action.indent),
                file=tangoktl.outf,
            )
        else:
            print(
                json.dumps(pod.to_dict(), indent=tangoktl.disp_action.indent, default=str),
                file=tangoktl.outf,
            )
        tangoktl.unset_output()
        return 0

    if tangoktl.disp_action.check(DispAction.TANGOCTL_NONE):
        tangoktl.disp_action.format = DispAction.TANGOCTL_DEFAULT
        _module_logger.info("Use default format %s", tangoktl.disp_action)

    tangoktl.domain_name = k8s.get_domain()
    if tangoktl.domain_name is None:
        _module_logger.error("Could not read domain name for context %s", k8s.context)
        return 1
    _module_logger.info("Domain name for context %s : %s", k8s.context, k8s.domain_name)
    tango_hosts: list[TangoHostInfo]
    tango_hosts = get_tango_hosts(
        _module_logger,
        tangoktl.tango_host,
        tangoktl.k8s_ns,
        tangoktl.cfg_data["databaseds_name"],
        f"svc.{tangoktl.domain_name}",
        tangoktl.cfg_data["databaseds_port"],
        tangoktl.use_fqdn,
        [],
    )
    if not tango_hosts:
        _ctx_name, _cluster_name, ns_list = tangoktl.get_namespaces_list()
        tango_hosts = get_tango_hosts(
            _module_logger,
            tangoktl.tango_host,
            tangoktl.k8s_ns,
            tangoktl.cfg_data["databaseds_name"],
            f"svc.{tangoktl.domain_name}",
            tangoktl.cfg_data["databaseds_port"],
            tangoktl.use_fqdn,
            ns_list,
        )
    if not tango_hosts:
        _module_logger.error("Could not read Tango hosts")
        return 1
    if len(tango_hosts) > 1:
        tangoktl.disp_action.quiet_mode = True
    _module_logger.info("Use Tango hosts %s", tango_hosts)

    if tangoktl.logging_level and tangoktl.tgo_name:
        return tangoktl.set_logging_level()

    thost: TangoHostInfo
    rc = 0
    ntango: int = 0
    ntangos: int = len(tango_hosts)
    for thost in tango_hosts:
        ntango += 1
        os.environ["TANGO_HOST"] = str(thost.tango_host)
        _module_logger.info("Set TANGO_HOST to %s", thost.tango_host)

        if tangoktl.disp_action.show_tango:
            print(f"TANGO_HOST={thost.tango_fqdn}:{thost.tango_port}")
            if thost.tango_ip is not None:
                print(f"TANGO_HOST={thost.tango_ip}:{thost.tango_port}")
            print()
            continue

        if tangoktl.disp_action.show_tree:
            verbose_tree: bool = False
            if tangoktl.disp_action.size == "L" or tangoktl.disp_action.size == "M":
                verbose_tree = True
            device_tree(include_dserver=tangoktl.disp_action.evrythng, verbose=verbose_tree)
            continue

        if tangoktl.input_file is not None:
            tangoktl.read_input_file()
            continue

        if tangoktl.dev_test:
            tangoktl.dev_ping = True
            tangoktl.dev_status = {"attributes": ["Status", "adminMode"]}
            tangoktl.disp_action.show_attrib = True
            tangoktl.disp_action.show_cmd = True
            tangoktl.disp_action.show_prop = True
        dev_test: bool = False
        if (
            tangoktl.dev_off
            or tangoktl.dev_on
            or tangoktl.dev_ping
            or tangoktl.dev_sim
            or tangoktl.dev_standby
        ):
            dev_test = True
        if tangoktl.dev_admin is not None:
            dev_test = True
        if dev_test and tangoktl.tgo_name:
            dut: TestTangoDevice = TestTangoDevice(_module_logger, tangoktl.tgo_name)
            if dut.dev is None:
                print(f"[FAILED] could not open device {tangoktl.tgo_name}")
                return 1
            # Run tests on Tango devices
            rc += dut.run_test(
                tangoktl.dev_admin,
                tangoktl.dev_off,
                tangoktl.dev_on,
                tangoktl.dev_ping,
                tangoktl.dev_sim,
                tangoktl.dev_standby,
                tangoktl.dev_status,
                tangoktl.disp_action.show_attrib,
                tangoktl.disp_action.show_cmd,
                tangoktl.disp_action.show_prop,
                tangoktl.tgo_attrib,
                tangoktl.tgo_name,
            )
            continue

        if tangoktl.disp_action.show_attrib:
            pass

        if tangoktl.tgo_value and tangoktl.tgo_attrib and tangoktl.tgo_name:
            rc = tangoktl.set_value()
            continue

        # Read info from Tango host
        rc += tangoktl.read_tango_host(ntango, ntangos)
    return rc


if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        pass
